examples/server/tests/features/server.feature

```diff
@@ -0,0 +1,148 @@
+Feature: llama.cpp server
+
+  Background: Server startup
+    Given a server listening on localhost:8080 with 2 slots, 42 as seed and llama.cpp as api key
+    Then  the server is starting
+    Then  the server is healthy
+
+  @llama.cpp
+  Scenario: Health
+    When the server is healthy
+    Then the server is ready
+    And  all slots are idle
+
+  @llama.cpp
+  Scenario Outline: Completion
+    Given a prompt <prompt>
+    And   a user api key <api_key>
+    And   <n_predict> max tokens to predict
+    And   a completion request
+    Then  <n_predict> tokens are predicted
+
+    Examples: Prompts
+      | prompt                           | n_predict | api_key   |
+      | I believe the meaning of life is | 128       | llama.cpp |
+      | Write a joke about AI            | 512       | llama.cpp |
+      | say goodbye                      | 0         |           |
+
+  @llama.cpp
+  Scenario Outline: OAI Compatibility
+    Given a system prompt <system_prompt>
+    And   a user prompt <user_prompt>
+    And   a model <model>
+    And   <max_tokens> max tokens to predict
+    And   streaming is <enable_streaming>
+    And   a user api key <api_key>
+    Given an OAI compatible chat completions request with an api error <api_error>
+    Then  <max_tokens> tokens are predicted
+
+    Examples: Prompts
+      | model        | system_prompt               | user_prompt                          | max_tokens | enable_streaming | api_key   | api_error |
+      | llama-2      | You are ChatGPT.            | Say hello.                           | 64         | false            | llama.cpp | none      |
+      | codellama70b | You are a coding assistant. | Write the fibonacci function in c++. | 512        | true             | llama.cpp | none      |
+      | John-Doe     | You are an hacker.          | Write segfault code in rust.         | 0          | true             | hackme    | raised    |
+
```

Comment:
@Azeirah As the server is started once for all tests, I have to include api_key in all completion requests. I did not manage to access openai request/request headers with that client version. So I cannot check the `Access-Control-Allow-Origin` is correctly set back.
Althought I think the last scenario `CORS Options` does what you were expected

Reply:
Oh the cors bug initially was caused by regardless of what the server had configured for api key. Some clients assume they're taking to real openAI server so they always send an api key, previous versions didn't handle that correctly.

For completeness sake though, the tests _should_ be able to restart the server I'd think, to verify correct operation with different parameters.

Perhaps it can be supervised by the python test script?

Reply:
yea good idea, I will change the `Background` steps to actually start the server, it will be more flexible for future scenario, allowing to test different server configs.